src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,170 @@
+    client_secret: str
+
+    @classmethod
+    def from_storage_permission_mapping(cls, storage_permission_mapping: StoragePermissionMapping, client_secret: str):
+        return cls(prefix=storage_permission_mapping.prefix,
+                   client_id=storage_permission_mapping.client_id,
+                   principal=storage_permission_mapping.principal,
+                   privilege=storage_permission_mapping.privilege,
+                   client_secret=client_secret)
+
+
+class AzureServicePrincipalMigration:
+
+    def __init__(self, installation: Installation, ws: WorkspaceClient, azure_resource_permissions: AzureResourcePermissions,
+                 azure_sp_crawler: AzureServicePrincipalCrawler):
+        self._final_sp_list = None
+        self._installation = installation
+        self._ws = ws
+        self._azure_resource_permissions = azure_resource_permissions
+        self._azure_sp_crawler = azure_sp_crawler
+        self._action_plan = 'service_principals_for_storage_credentials.csv'
+
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, customized_csv: str, replace_with_ac: bool, product='ucx'):
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        azure_resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        azure_sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        return cls(installation, ws, azure_resource_permissions, azure_sp_crawler, customized_csv, replace_with_ac)
+
+
+    def _list_storage_credentials(self):
+        # list existed storage credentials that is using service principal, capture the service principal's application_id
+        return {}
+
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set):
+        # if sp is already used, take it off from the sp_list
+        return list()
+
+
+    def _fetch_client_secret(self, sp_list: list[StoragePermissionMapping]) -> list[ServicePrincipalMigrationInfo]:
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the client_secret field
+        #
+        # The input StoragePermissionMapping may have managed identity mixed in, we will ignore them for now, as
+        # they won't have any client_secret, we will process managed identity in the future.
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if azure_sp_info.secret_scope is None:
```

Comment:
Invert the logic, this method is too long: given the spn info instance, retrieve a secret from it and return as string

Reply:
split it into two functions now